File: Engine/Android/core/src/main/kotlin/despia/engine/ChainResolver.kt
//
//  ChainResolver.kt - module-identity CHAIN resolution (the chains corpus law).
//
//  A module's identity is its dotted chain derived from Modules/ nesting (`watch.health`);
//  the chain is BOTH the API face (`dsx.module.watch.health.heartRate`) and the wire scheme
//  token (`watch.health://heartRate`). Resolution is NEVER dot-counting and never a first-dot
//  split — it is the incremental FOLD at each runtime's ONE dispatch funnel:
//
//    1. normalize the arriving HEAD spelling through the alias map to its primary chain
//       (`watchhealth` → `watch.health`; an unknown head stays itself, `known = false` — so
//       module_not_found reports against the head token, never a silent no-op);
//    2. while the remaining action path is non-empty and `chain + "." + parts[0]` is in the
//       IDENTITY SET — registered chains ∪ build-excluded chains, the honest universe, so an
//       excluded child still gets correct attribution instead of becoming a phantom action on
//       its parent — fold that segment into the chain;
//    3. what remains is the action path (empty = the bare pre-filter call).
//
//  The build-time bidirectional ban (a child segment can never equal a parent action/group
//  first-segment or a reserved member — dsx_graph chain_errors) is what makes the fold total
//  and unambiguous: `intelligence.rag.add` can only ever mean the `rag` GROUP, because a child
//  named `rag` could not have built. Aliases are consulted at HEAD position only (they route
//  legacy wire tokens — hyphens fine — never the modern dot plane); the arriving spelling stays
//  visible to the module (`ctx.scheme` — catch-all routing) while identity (state, events,
//  config) is always the primary chain.
//
//  RESERVED MEMBERS are never actions: after the fold, a leading reserved word in the
//  remainder routes to the MEMBER plane — the call parser refuses it as an action. One that
//  nevertheless arrives at a MODERN call face — the typed proxies, the dotted-callee face,
//  the native _call/_dispatch funnel (impossible through the proxies themselves — a caller
//  bug) — is REFUSED with the cross-runtime error code `reserved_member`: reported through
//  the module.callFailed funnel, thrown as the runtime's ModuleCallError, never dispatched
//  as an action and never a silent unknown_action (frozen in the corpus `_note`; the TS bus
//  and Swift Context.foldRoute spell it identically). The LEGACY WIRE face (URL navigations,
//  the v3 string transport) is EXEMPT from the refusal: reserved words are banned from
//  MANIFESTS, never from the wire, so a code-only legacy shim registered under a reserved
//  spelling (`biometric://available`, `bluetooth://state`, `scanningmode://on`) keeps
//  answering shipped pages — the compat those shims exist for. On the wire face the FOLD
//  still applies for ROUTING — a nested chain reaches its owner even when the remainder
//  starts with a reserved word (`chfx://sub/state` dispatches action `state` on `chfx.sub`) —
//  so owner and attribution are the same on every runtime; only the refusal is face-gated.
//  The set is
//  CLOSED and FROZEN (growth is a major-version event): native runtimes are proxy-safe by
//  construction (reserved members are REAL members, and real members shadow dynamic lookup),
//  so the JS planes' proxy denylist (`then`/`toString`/…) has no twin here.
//
//  Pure JVM, dependency-free, unit-testable: we take an identity VIEW (an interface), so the
//  same fold runs against the corpus's synthetic table (ChainsConformanceTest) AND the live
//  route table (ModuleRegistry adapts itself). The kernel names no modules — every chain in
//  here arrives as data. Corpus: OpenSource/Conformance/chains/chains.json (TS + Kotlin
//  execute every case; Swift is the reference lane).
//

package despia.engine

object ChainResolver {

    /// The FROZEN member plane — never actions, closed set (growth = major-version event).
    /// Mirrors `chains.json` `reserved`; the corpus runner asserts the two stay identical.
    val reservedMembers: Set<String> = setOf(
        "on", "available", "excluded", "state", "context", "object", "delegate", "dsx", "then")

    /// The identity view the fold runs against — the resolver never owns the table, it asks.
    /// Implementations: [Table] (the corpus's synthetic table, exact-match) and the
    /// ModuleRegistry adapter (live routes ∪ the build's excluded overlay, keyed lowercase).
    interface Identity {
        /// The primary chain `head` is an ALIAS of, or null when `head` is not an alias
        /// (i.e. it is already a primary spelling, or unknown). Consulted at HEAD position only.
        fun primaryFor(head: String): String?
        /// Membership in the IDENTITY SET: registered chains ∪ build-excluded chains.
        fun contains(chain: String): Boolean
        /// True when `chain` resolves through the excluded overlay (build fact, not a route).
        fun isExcluded(chain: String): Boolean
    }

    /// The corpus-shaped identity table: chains + alias→chain + the excluded overlay.
    class Table(
        private val chains: Set<String>,
        private val aliases: Map<String, String> = emptyMap(),
        private val excluded: Set<String> = emptySet(),
    ) : Identity {
        override fun primaryFor(head: String): String? = aliases[head]
        override fun contains(chain: String): Boolean = chain in chains || chain in excluded
        override fun isExcluded(chain: String): Boolean = chain in excluded
    }

    /// One resolved callee. Exactly the corpus's expectation shape: a CALL route carries
    /// `action` (member null); a MEMBER route carries `member` + `rest` (action empty — the
    /// call parser refuses a reserved word as an action, so we never expose it as one).
    class Resolution(
        /// The resolved primary identity — the dotted chain (state, events, config key).
        val chain: String,
        /// The action path segments after the fold ("" segments never occur; empty list = the
        /// bare pre-filter call). Empty when this is a member route.
        val action: List<String>,
        /// The ARRIVING spelling of the chain — equals `chain` except when the head arrived as
        /// an alias (`watchhealth`, `get-uuid`): legacy grammar, kept visible for catch-all
        /// routing (`ctx.scheme`) while identity stays the primary chain.
        val spelling: String,
        /// False when the head resolved into nothing we know — the honest module_not_found
        /// attribution target is then `chain` (the head token), never a silent no-op.
        val known: Boolean,
        /// True when identity resolved through the excluded overlay — the module is real but
        /// not in this build; attribution and build facts (`available`/`excluded`) still answer.
        val excluded: Boolean,
        /// The reserved member word leading the post-fold remainder, or null for a call route.
        val member: String?,
        /// The member-plane tail after the member word (empty for call routes).
        val rest: List<String>,
        /// How many action-path segments the fold moved into the chain (0 = head-only). Lets a
        /// dispatch adapter fall back to its verbatim pre-fold behavior when nothing folded.
        val folded: Int,
    )

    /// Step 2 alone — the incremental fold, shared verbatim by [resolve] and the dispatch
    /// funnel (which walks a separator-preserving cursor instead of a pre-split list): from
    /// `base`, consume leading `tokens` while `chain + "." + token` is in the identity set.
    /// Returns how many tokens folded. An empty token never folds (malformed path — leave it
    /// to the action plane's own error answer).
    fun fold(base: String, tokens: List<String>, identity: Identity): Int {
        var chain = base
        var n = 0
        while (n < tokens.size) {
            val token = tokens[n]
            if (token.isEmpty()) break
            val next = "$chain.$token"
            if (!identity.contains(next)) break
            chain = next
            n += 1
        }
        return n
    }

    /// Resolve one dotted callee (the segments after `dsx.module.`; a single segment = the
    /// bare scheme-only call) against `identity`. Pure — no registry, no dispatch, no I/O.
    fun resolve(segments: List<String>, identity: Identity): Resolution {
        val head = segments.firstOrNull() ?: ""
        // (1) alias-normalize the HEAD — and only the head — to its primary chain.
        val primary = identity.primaryFor(head)
        var chain = primary ?: head
        var spelling = head
        // (2) the fold: longest KNOWN prefix wins, one segment at a time.
        val tail = segments.drop(1)
        val folded = fold(chain, tail, identity)
        for (i in 0 until folded) {
            chain = "$chain.${tail[i]}"
            spelling = "$spelling.${tail[i]}"
        }
        // (3) the remainder — a leading reserved word routes to the MEMBER plane, never an action.
        val remainder = tail.drop(folded)
        val member = remainder.firstOrNull()?.takeIf { it in reservedMembers }
        return Resolution(
            chain = chain,
            action = if (member == null) remainder else emptyList(),
            spelling = spelling,
            known = identity.contains(chain),
            excluded = identity.isExcluded(chain),
            member = member,
            rest = if (member == null) emptyList() else remainder.drop(1),
            folded = folded)
    }
}

File: Engine/iOS/ChainResolver.swift
//
//  ChainResolver.swift — derived module identity: the longest-known-prefix FOLD.
//
//  A module's identity is its dotted CHAIN (watch.health), derived at build time from
//  `Modules/` nesting (facet-contracts.md) — the manifest declares only its LOCAL segment.
//  The chain is BOTH the API face (dsx.module.watch.health.heartRate) and the wire scheme
//  token (watch.health://heartRate). This file is the REFERENCE implementation of the
//  resolution law executed by OpenSource/Conformance/chains/chains.json on all three
//  runtimes (TS vitest · Kotlin :core JUnit · Swift, this file):
//
//    1. Normalize the arriving HEAD spelling through the alias map to its primary chain
//       (watchhealth → watch.health). An unknown head stays itself (known = false) — the
//       module_not_found path reports against the head token, never a silent no-op.
//    2. While the remaining path is non-empty and `chain + "." + parts[0]` is in the
//       IDENTITY SET (registered chains ∪ build-excluded chains — the honest universe, so
//       an excluded child still gets correct attribution instead of becoming a phantom
//       action on its parent), fold that segment into the chain.
//    3. What remains is the action path (slash-joined, the array-path convention; empty =
//       the bare pre-filter call) — UNLESS its first token is a reserved member, which
//       routes to the MEMBER plane and is never an action.
//
//  The build-time bidirectional ban (dsx_graph chain_errors: a child segment can never
//  equal a parent action/group first-segment or a reserved member, and vice versa) is what
//  makes the fold total and unambiguous — intelligence.rag.add can only ever mean the rag
//  GROUP, because a child named rag could not have built.
//
//  The reserved set is CLOSED and FROZEN: these are REAL members on the typed proxies,
//  and real members shadow dynamic lookup (SE-0195) — a new word could silently steal a
//  module's action name, so growth is a major-version event, never a patch.
//

import Foundation

public enum ChainResolver {

    /// The module-proxy reserved members — the frozen nine (dsx_graph RESERVED_MEMBERS twin).
    public static let reservedMembers: Set<String> =
        ["on", "available", "excluded", "state", "context", "object", "delegate", "dsx", "then"]

    /// The identity view the fold runs against — built by ModuleRegistry from live
    /// registrations plus the build's excluded overlay (never hand-assembled by callers).
    public struct Table {
        /// Primary identity chains (registered modules).
        public let chains: Set<String>
        /// Legacy spelling → primary chain (aliases route at HEAD position only).
        public let aliases: [String: String]
        /// Build-excluded identities (chains that exist in the product but not this build).
        public let excluded: Set<String>
        /// The honest universe: everything that IS an identity, shipped or not —
        /// PRECOMPUTED at construction (the fold probes it once per segment; a
        /// computed union would allocate O(catalog) per probe).
        let identity: Set<String>
        /// chain → its legacy alias spellings (the reverse of `aliases`, sorted for
        /// determinism) — the emission fan-out reads this so an alias handle hears
        /// a module's events no matter when it subscribed.
        public let aliasesByChain: [String: [String]]
        public init(chains: Set<String>, aliases: [String: String] = [:], excluded: Set<String> = []) {
            self.chains = chains
            self.aliases = aliases
            self.excluded = excluded
            self.identity = chains.union(excluded)
            var reverse: [String: [String]] = [:]
            for (alias, chain) in aliases { reverse[chain, default: []].append(alias) }
            self.aliasesByChain = reverse.mapValues { $0.sorted() }
        }
    }

    public struct Resolution {
        /// The resolved PRIMARY identity (or the arriving head itself when unknown).
        public let chain: String
        /// Slash-joined action path ("" = the bare pre-filter call). Empty when `member`
        /// routes. ADVISORY spelling only — a funnel that must preserve the caller's
        /// grammar (dotted group keys, case) cuts the ORIGINAL string with
        /// `consume(_:folded:)` instead.
        public let action: String
        /// The arriving spelling of the folded chain. This equals the caller's
        /// dotted chain for primary names; a legacy alias remains its one head token.
        public let spelling: String
        /// False when the head names no known identity (module_not_found attribution).
        public let known: Bool
        /// True when the resolved identity sits under the build's excluded overlay.
        public let excluded: Bool
        /// A reserved member routed instead of an action (never dispatched as a call
        /// on the MODERN faces; the wire face dispatches it as a plain action).
        public let member: String?
        /// The dot-joined remainder after a member route ("bpm" in …state.bpm).
        public let rest: String
        /// How many ACTION segments the fold consumed into the chain (0 = nothing
        /// folded — the caller keeps its verbatim route, bit-for-bit).
        public let folded: Int
    }

    /// Step 2 alone — the incremental fold shared by `resolve` (the corpus's pre-split
    /// reference face) and `resolveWire` (the wire funnel), the Kotlin
    /// `ChainResolver.fold` twin: from `base`, consume leading `tokens` while
    /// `chain + "." + token` is in the identity set; returns how many folded. Tokens
    /// arrive LOWERCASED. An EMPTY token never folds (a malformed path stays the
    /// action plane's own error answer, never a half-folded route).
    static func fold(base: String, tokens: [String], table: Table) -> Int {
        var chain = base
        var n = 0
        while n < tokens.count {
            let token = tokens[n]
            if token.isEmpty { break }
            let next = "\(chain).\(token)"
            guard table.identity.contains(next) else { break }
            chain = next
            n += 1
        }
        return n
    }

    /// Resolve dotted-callee segments ([head, part, part…]) against the table. Segments
    /// arrive pre-split — callers split their surface's spelling ("/" and "." are both
    /// path separators on the wire; a segment name can contain neither). Tokens compare
    /// LOWERCASED against the lowercase identity table (twin parity: Kotlin lowercases
    /// before its fold, TS folds case-insensitively); the membership check is the ONLY
    /// oracle — folding never depends on the head's own standing.
    public static func resolve(_ segments: [String], table: Table) -> Resolution {
        guard let head = segments.first, !head.isEmpty else {
            return Resolution(chain: "", action: "", spelling: "", known: false,
                              excluded: false, member: nil, rest: "", folded: 0)
        }
        var spelling = head
        let headKey = head.lowercased()
        let alias = table.aliases[headKey]
        var chain = alias ?? headKey
        let tokens = segments.dropFirst().map { $0.lowercased() }
        let folded = fold(base: chain, tokens: tokens, table: table)
        for i in 0..<folded {
            chain += ".\(tokens[i])"
            // A dotted primary call exposes the complete arriving spelling to
            // ctx.scheme. A legacy alias is one indivisible wire token.
            if alias == nil { spelling += ".\(segments[i + 1])" }
        }
        let known = table.identity.contains(chain)
        let parts = Array(segments.dropFirst(1 + folded))
        let excluded = table.excluded.contains(chain)
            || table.excluded.contains(where: { chain.hasPrefix("\($0).") })
        if let first = parts.first, reservedMembers.contains(first.lowercased()) {
            return Resolution(chain: chain, action: "", spelling: spelling, known: known,
                              excluded: excluded, member: first.lowercased(),
                              rest: parts.dropFirst().joined(separator: "."), folded: folded)
        }
        return Resolution(chain: chain, action: parts.joined(separator: "/"),
                          spelling: spelling, known: known, excluded: excluded,
                          member: nil, rest: "", folded: folded)
    }

    /// One resolved WIRE callee — the Kotlin `ModuleRegistry.resolveChain` twin,
    /// sized for the string/URL funnels: `chain` the folded primary identity,
    /// `rest` the action-path remainder cut VERBATIM from the original (separators
    /// and case preserved, so dotted group keys keep their registration grammar),
    /// `folded` how many segments the fold consumed (0 ⇒ the caller keeps its
    /// bit-for-bit pre-fold route), `member` the reserved word leading the
    /// post-fold remainder (lowercased), when there is one.
    public struct WireResolution {
        public let chain: String
        public let rest: String
        public let folded: Int
        public let member: String?
    }

    /// THE funnel-shaped fold over a wire spelling — every Swift fold site
    /// (`handle(url:)`, the string funnel, `foldRoute`, `normalizeCall`) resolves
    /// through THIS one helper, byte-for-byte the Kotlin funnel's law:
    /// strip ONE leading "/" (the array-path spelling), tokenize on "/" and "."
    /// KEEPING empty tokens with each token's past-separator boundary recorded,
    /// alias-normalize the head, fold while the identity set knows the deeper
    /// chain — an EMPTY token never folds (a malformed path stays the action
    /// plane's own error answer, never a half-folded route) — then cut the
    /// remainder VERBATIM at the recorded boundary. Counting separator
    /// CHARACTERS instead desyncs from an empty-omitting split on doubled or
    /// leading separators and re-dispatches a folded segment (review round 3).
    /// `dsx` is the reserved kernel channel, not a module — it never folds.
    public static func resolveWire(scheme: String, actionPath: String, table: Table) -> WireResolution {
        let key = scheme.lowercased()
        let path = actionPath.hasPrefix("/") ? String(actionPath.dropFirst()) : actionPath
        if key.isEmpty || key == "dsx" { return WireResolution(chain: key, rest: path, folded: 0, member: nil) }
        var tokens: [String] = []
        var after: [String.Index] = []     // index PAST token i's separator (== endIndex at the tail)
        var start = path.startIndex
        while start < path.endIndex {
            var cut = start
            while cut < path.endIndex, path[cut] != "/", path[cut] != "." { cut = path.index(after: cut) }
            tokens.append(path[start..<cut].lowercased())
            let next = cut < path.endIndex ? path.index(after: cut) : cut
            after.append(next)
            start = next
        }
        var chain = table.aliases[key] ?? key
        var folded = 0
        while folded < tokens.count {
            let token = tokens[folded]
            if token.isEmpty { break }
            let next = "\(chain).\(token)"
            guard table.identity.contains(next) else { break }
            chain = next
            folded += 1
        }
        let rest = folded == 0 ? path : String(path[after[folded - 1]...])
        let member = folded < tokens.count && reservedMembers.contains(tokens[folded]) ? tokens[folded] : nil
        return WireResolution(chain: chain, rest: rest, folded: folded, member: member)
    }
}
